### ISSUE TITLE
apps deployed to Homebrew tomcat are lost after brew update

### ISSUE THREAD

**Issue by ryenus** [Homebrew/homebrew-core#43311]
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [x] are reporting a bug others will be able to reproduce and not asking a question or requesting software. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh. To get software added or changed in Homebrew please file a [Pull Request](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)
- [x] have a problem with `brew install` (or `upgrade`, `reinstall`) a single, Homebrew/homebrew-core formula (not cask) on macOS? If it's a general `brew` problem please file this issue at Homebrew/brew: https://github.com/Homebrew/brew/issues/new/choose. If it's a Linux problem please file this issue at https://github.com/Homebrew/linuxbrew-core/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [x] ran `brew update` and can still reproduce the problem?
- [x] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [x] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

To help us debug your issue please explain:

- What you were trying to do (and why)
  I have deployed some web apps to tomcat, and I want those web apps are migrated to the updated tomcat installation after brew updating tomcat.

- What happened (include command output)
  By default the web apps are deployed to `/usr/local/opt/tomcat/libexec/webapps`, after brew updating tomcat, `/usr/local/opt/tomcat` now points to the newer version, that basically throw the previously deployed web apps away.

- What you expected to happen
  The web apps deployed to previous installed tomcat versions should be migrated to the latest.

- Step-by-step reproduction instructions (by running `brew install` commands)
  1. brew install tomcat
  2. deploy any web app to tomcat, or make any change to the builtin apps
  3. update tomcat, or simply uninstall, then reinstall.
  4. you'd see that any change to the apps in `/usr/local/opt/tomcat/libexec/webapps` are now lost, replaced by the default apps shipped with latest tomcat.

<details>
<summary>brew config output</summary>
<table>
<tr><td>HOMEBREW_VERSION</td> <td>2.1.10-30-g84c5f40</td></tr>
<tr><td>ORIGIN</td> <td>https://github.com/Homebrew/brew.git</td></tr>
<tr><td>HEAD</td> <td>84c5f4079347fb7bc7844d594f5735347abf5607</td></tr>
<tr><td>Last commit</td> <td>2 hours ago</td></tr>
<tr><td>Core tap ORIGIN</td> <td>https://github.com/Homebrew/homebrew-core</td></tr>
<tr><td>Core tap HEAD</td> <td>ea3402811b5fed36d56f259b93e5e70fb5dc4d53</td></tr>
<tr><td>Core tap last commit</td> <td>12 hours ago</td></tr>
<tr><td>HOMEBREW_PREFIX</td> <td>/usr/local</td></tr>
<tr><td>HOMEBREW_DEV_CMD_RUN</td> <td>1</td></tr>
<tr><td>HOMEBREW_GITHUB_API_TOKEN</td> <td>set</td></tr>
<tr><td>HOMEBREW_NO_ANALYTICS</td> <td>1</td></tr>
<tr><td>CPU</td> <td>dodeca-core 64-bit kabylake</td></tr>
<tr><td>Homebrew Ruby</td> <td>2.3.7 => /System/Library/Frameworks/Ruby.framework/Versions/2.3/usr/bin/ruby</td></tr>
<tr><td>Clang</td> <td>10.0 build 1001</td></tr>
<tr><td>Git</td> <td>2.23.0 => /usr/local/bin/git</td></tr>
<tr><td>Curl</td> <td>7.54.0 => /usr/bin/curl</td></tr>
<tr><td>Java</td> <td>12.0.2, 11.0.4, 1.8.0_222</td></tr>
<tr><td>macOS</td> <td>10.14.6-x86_64</td></tr>
<tr><td>CLT</td> <td>10.3.0.0.1.1562985497</td></tr>
<tr><td>Xcode</td> <td>N/A</td></tr>
<tr><td>CLT headers</td> <td>10.3.0.0.1.1562985497</td></tr>
</table>
</details>

<details>
<summary>brew doctor output</summary>
<pre>
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Putting non-prefixed coreutils in your path can cause gmp builds to fail.
</pre>
</details>

**Comment by ryenus** [comment]
tl;dr

There're 2 most significant environment variables when running tomcat:
* `$CATALINA_HOME`: namely the tomcat installation dir
* `$CATALINA_BASE`:  to store runtime files, defaults to `CATALINA_HOME`, but can be different

One possible solution is to leverage `$CATALINE_BASE` and point it to a directory that can be kept intact when brew updating tomcat, e.g. `/usr/local/var/tomcat`. 

And for this to work we have to make sure the '$CATALINA_BASE` directory is defined, and make sure it's populated with necessary content upon first run. We can do this by:
1. replace the symbol link `/usr/local/opt/tomcat/bin/catalina`, with a wrapper script, and within this script, it does the following:
2. `export CATALINA_BASE=$(brew --prefix)/var/tomcat`
3. copy tomcat/conf into `$CATALINA_BASE`
4. create context files for the OOTB web apps in ``$CATALINA_BASE/conf/Catalina/localhost/`
5. call `/usr/local/opt/tomcat/libexec/bin/catalina.sh`

For your reference, this is something I've come up with:
https://gist.github.com/ryenus/d4aaac85655e62165cf58acbfda55b59

And the related documentation can be found at:
https://tomcat.apache.org/tomcat-9.0-doc/RUNNING.txt

```
==================================================
Advanced Configuration - Multiple Tomcat Instances
==================================================

In many circumstances, it is desirable to have a single copy of a Tomcat
binary distribution shared among multiple users on the same server.  To make
this possible, you can set the CATALINA_BASE environment variable to the
directory that contains the files for your 'personal' Tomcat instance.
```

**Comment by iMichka** [comment]
Hi. It looks like you found a workaround. We do not have the manpower and resources to help you on this. It looks like you made a first attempt to fix this with a pull request: we would review another one if you find a cleaner solution.

I'll close this issue as we will not be able to move this forward.
